### PR TITLE
Allow multiple tokens for single name

### LINKED ISSUE #16: Wildcard SAN issue
Just a heads up, there seems to be an issue with certificates that contain the wildcard name as a SAN.

First reported issue: https://github.com/lukas2511/dehydrated/issues/420#issuecomment-372954837
More detailed issue: https://github.com/lukas2511/dehydrated/issues/504

### PR DESCRIPTION
**Pull request by qvr**
Wildcard + apex certificates need two challenges, both for the same "name". With this PR we first save all unique names and their respective token(s), and then create the needed rrsets with multiple records if needed.

I haven't tested this extensively, but at least "*.domain.tld domain.tld foo.tld" certs work.

HOOK_CHAIN=yes is of course required to get all the tokens for a name at the same time (and I couldn't get the timing to work without HOOK_CHAIN without this PR).

I also moved the wildcard check from setup_domain to main, but as far as I can tell dehydrated won't ever send names like that (*.domain.com) so I guess that commit could be reverted altogether.

This should fix #16

### CODE REVIEW

**Comment by silkeh** [comment]
Thanks for this! This is almost exactly what I had in mind, barring the given comments.

File: pdns_api.sh [hunk 1]
```diff
@@ -241,16 +236,25 @@ setup_domain() {
...
-  echo '{
+  echo -n '{
     "name": "'"${name}"'",
     "type": "TXT",
     "ttl": 1,
-    "records": [{
-      "content": "\"'"${token}"'\"",
+    "records": ['
+
+    tokens=( ${token} )
```
[the remark this answers is not in the document]
Comment (qvr) [reply pdns_api.sh]: Changed to read -a

File: pdns_api.sh [hunk 2]
```diff
@@ -241,16 +236,25 @@ setup_domain() {
...
 
 deploy_rrset() {
-  echo '{
+  echo -n '{
     "name": "'"${name}"'",
     "type": "TXT",
     "ttl": 1,
-    "records": [{
-      "content": "\"'"${token}"'\"",
+    "records": ['
```
[the remark this answers is not in the document]
Comment (qvr) [reply pdns_api.sh]: Done, using get_records

File: pdns_api.sh [hunk 3]
```diff
@@ -343,16 +347,31 @@ main() {
...
     req=""
-    t=$((i + 2))
-    setup_domain "${!i}" "${!t}"
+    t=${domains[${domain}]}
+    setup_domain "${domain}" "${t}"
 
     # Debug output
     debug "Name:  ${name}"
-    debug "Token: ${token}"
+    debug "Token(s): ${token}"
```
[the remark this answers is not in the document]
Comment (qvr) [reply pdns_api.sh]: Removed the change

File: pdns_api.sh [hunk 4]
```diff
@@ -241,16 +236,25 @@ setup_domain() {
...
     "name": "'"${name}"'",
     "type": "TXT",
     "ttl": 1,
-    "records": [{
-      "content": "\"'"${token}"'\"",
+    "records": ['
+
+    tokens=( ${token} )
+    for i in "${!tokens[@]}"; do
+      printf '%.*s' $((i != 0)) ","
```
[the remark this answers is not in the document]
Comment (qvr) [reply pdns_api.sh]: I think most of my time on this PR was already spent on that comma... :-)

I couldn't figure out how to not depend on the index (or a separate iteration counter, which was more useless lines).

**Comment by jangrewe** [comment]
This works fine for me! Any reason against merging it? ;-)

File: pdns_api.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (silkeh) [reply pdns_api.sh]: I think the other solution would be to append a string, but that's not that elegant either. Let's keep it for now.

**Comment by silkeh** [comment]
LGTM. Thanks!

**Pull request merged**